runners/core-java/src/main/java/org/apache/beam/runners/core/SideInputHandler.java

```diff
@@ -183,8 +177,9 @@ public void addSideInputValue(PCollectionView<?> sideInput, WindowedValue<Iterab
     ValueState<Iterable<?>> state =
         stateInternals.state(StateNamespaces.window(windowCoder, window), stateTag);
 
-    // returns null when the side input is not ready
-    return state.read();
+    Iterable<?> elements = state.read();
+    // return empty collection when no side input was received for ready window
+    return (elements != null) ? elements : Collections.emptyList();
```

Comment:
Slightly suspicious that we don't receive elements here. I understand that side inputs may not always be ready depending on the trigger strategy. We discovered the bug running Wordcount in streaming mode, which uses side inputs for the write operations. The trigger there is the default trigger, i.e. end of global window. Shouldn't all side inputs be ready then?

Reply:
I think it is `pre_finalize_coll` in `iobase.py`. That's a flat map which indeed does not output any data. It would be nice to be able to distinguish "not ready" from "no output". I suppose if the key exists in the state backend, we can assume that (which is in lines with this PR).

Reply:
There is indeed a mystery how a window can become "ready" without having added a value. (The only call that mutates `availableWindows` in `SideInputHandler` is `addSideInputValue`. That's how the old runner already works however.